lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -0,0 +1,169 @@
+
+//===----------------------------------------------------------------------===//
+// Pass driver
+//===----------------------------------------------------------------------===//
+class SCFToCalyxPass : public SCFToCalyxBase<SCFToCalyxPass> {
+public:
+  SCFToCalyxPass() : SCFToCalyxBase<SCFToCalyxPass>() {}
+  void runOnOperation() override;
+
+  LogicalResult mainFuncIsDefined(mlir::ModuleOp moduleOp,
+                                  StringRef topLevelFunction) {
+    if (SymbolTable::lookupSymbolIn(moduleOp, topLevelFunction) == nullptr) {
+      moduleOp.emitError("Main function '" + topLevelFunction +
+                         "' not found in module.");
+      return failure();
+    }
+    return success();
+  }
+
+  //// Creates a new Calyx program with the contents of the source module
+  /// inlined within.
+  /// Furthermore, this function performs validation on the input function, to
+  /// ensure that we've implemented the capabilities necessary to convert it.
+  ///
+  /// @todo: this seems unnecessarily complicated:
+  /// A restriction of the current infrastructure is that a top-level 'module'
+  /// cannot be overwritten (even though this is essentially what is going on
+  /// when replacing standard::ModuleOp with calyx::ProgramOp). see:
+  /// https://llvm.discourse.group/t/de-privileging-moduleop-in-translation-apis/3733/26
+  LogicalResult createProgram(calyx::ProgramOp *programOpOut) {
+    // Program conversion
+    auto createModuleConvTarget = [&]() {
+      ConversionTarget target(getContext());
+      target.addLegalDialect<calyx::CalyxDialect>();
+      target.addLegalDialect<scf::SCFDialect>();
+      target.addIllegalDialect<hw::HWDialect>();
+      target.addIllegalDialect<comb::CombDialect>();
+
+      // For loops should have been lowered to while loops
+      target.addIllegalOp<scf::ForOp>();
+
+      // Only accept std operations which we've added lowerings for
+      target.addIllegalDialect<StandardOpsDialect>();
+      target.addLegalOp<AddIOp, SubIOp, CmpIOp, ShiftLeftOp,
+                        UnsignedShiftRightOp, SignedShiftRightOp, AndOp, XOrOp,
+                        OrOp, ZeroExtendIOp, TruncateIOp, CondBranchOp,
+                        BranchOp, ReturnOp, ConstantOp, IndexCastOp>();
+
+      target.addDynamicallyLegalOp<mlir::ModuleOp>([](mlir::ModuleOp moduleOp) {
+        // A module is legalized after we've added a nested
+        // calyx::ProgramOp within it.
+        bool ok = false;
+        moduleOp.walk([&](calyx::ProgramOp) {
+          ok = true;
+          return WalkResult::interrupt();
+        });
+        return ok;
+      });
+      return target;
+    };
+
+    RewritePatternSet patterns(&getContext());
+    patterns.add<ModuleOpConversion>(&getContext(), programOpOut);
+    auto target = createModuleConvTarget();
+    return applyPartialConversion(getOperation(), target, std::move(patterns));
+  }
+
+private:
+  mlir::ModuleOp m_moduleOp;
+  calyx::ProgramOp m_programOp;
```

Comment:
Does this state abide by the restrictions here? https://mlir.llvm.org/docs/PassManagement/#operation-pass

More generally, what is this needed for? My understanding is to force the pass to only support converting one module in the input IR? Could we just remove this and the checks based on it? Generally, I'd expect 1 module in the input to this pass, but if there are N modules, is there a problem converting each one to its own program?

Reply:
The `m_programOp` will be used to initialize the `ProgramLoweringState` of #1814. There is no real need to keep it as a member variable, so i'll modify that to a temporary variable.

Based on the criteria of the referenced site, the pass does abide by what is laid out there.